extractor/main_test.go

```diff
@@ -0,0 +1,412 @@
+		require.NoError(t, err)
+
+		reader := bufio.NewReader(file)
+		var unmarshalledMetadata Metadata
+
+		raw, isPrefix, err := reader.ReadLine()
+		require.NoError(t, err)
+
+		if isPrefix {
+			t.Fatalf("Metadata longer then buffer max size at %s\n", outputDir)
+		}
+
+		err = json.Unmarshal(raw, &unmarshalledMetadata)
+		require.NoError(t, err)
+
+		t.Cleanup(func() {
+			err := os.RemoveAll(tempDir)
+			if err != nil {
+				fmt.Printf("could not clean up temp dir, %v", err)
+				return
+			}
+		})
+
+		require.Equal(t, md, unmarshalledMetadata)
+	}
+}
+
+func TestWritePackageFiles(t *testing.T) {
+	t.Parallel()
+
+	_, mockMsgsAddPackage := generateMockMsgs(t)
+
+	tempDir, err := os.MkdirTemp(".", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	for _, msg := range mockMsgsAddPackage {
+		// Get output dir
+		outputDir := filepath.Join(tempDir, strings.TrimLeft(msg.Package.Path, "gno.land/"))
+
+		// Write dir before writing metadata
+		if dirWriteErr := os.MkdirAll(outputDir, os.ModePerm); dirWriteErr != nil {
+			t.Fatal(dirWriteErr)
+		}
+
+		// Write the metadata
+		err := writePackageFiles(msg, outputDir)
+		require.NoError(t, err)
+
+		// Read & compare file
+		for _, f := range msg.Package.Files {
+			contents, err := os.ReadFile(filepath.Join(outputDir, f.Name))
+			require.NoError(t, err)
+			require.Equal(t, f.Body, string(contents))
+		}
+	}
+	t.Cleanup(func() {
+		err := os.RemoveAll(tempDir)
+		if err != nil {
+			fmt.Printf("could not clean up temp dir, %v", err)
+			return
+		}
+	})
+}
+
+// Helpers
+func generateSourceFiles(t *testing.T, mockMsgs []std.Msg) []string {
+	t.Helper()
+
+	tempDir, err := os.MkdirTemp(".", "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var (
+		mockTx    = make([]std.Tx, numTx)
+		testFiles = make([]string, numSourceFiles)
+	)
+
+	// Generate transactions to wrap messages
+	for i := range mockTx { // num
+		mockTx[i] = std.Tx{
+			Msgs: mockMsgs[:msgPerTx],
+		}
+		mockMsgs = mockMsgs[msgPerTx:]
+	}
+
+	// Generate source file names
+	for i := 0; i < numSourceFiles; i++ {
+		testFiles[i] = "sourceFile" + strconv.Itoa(i) + sourceFileType
+	}
+
+	// Generate source files
+	for _, file := range testFiles {
+		filePath := filepath.Join(tempDir, file)
+		if err := os.MkdirAll(filepath.Dir(filePath), os.ModePerm); err != nil {
+			t.Fatal(err)
+		}
+		file, err := os.Create(filePath)
+
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		for _, tx := range mockTx[:txPerSourceFile] {
+
+			err := writeTxToFile(tx, file, t)
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+		mockTx = mockTx[txPerSourceFile:]
+	}
+
+	for i := 0; i < numSourceFiles; i++ {
+		testFiles[i] = filepath.Join(tempDir, testFiles[i])
+	}
+
+	t.Cleanup(func() {
+		err := os.RemoveAll(tempDir)
+		if err != nil {
+			fmt.Printf("could not clean up temp dir, %v", err)
+			return
+		}
+	})
+
+	return testFiles
+}
+
+func generateMockMsgs(t *testing.T) ([]std.Msg, []vm.MsgAddPackage) {
```

Comment:
Why do you need these mocks to be as detailed as you make them here?

You are using these mock messages to filter and verify your parse methods work, there is no need to have them be this intricately set up

Reply:
That's how I initially made them, I don't think it's worth the effort to refactor this as it works and it's readable.

Reply:
> That's how I initially made them, I don't think it's worth the effort to refactor this as it works and it's readable.

I'm sorry, but I don't understand the reasoning "it's not worth the effort to refactor". The whole point of this review process is to get feedback and apply it if it makes sense. The logic you wrote here is way too complicated for what it needs to be doing.

Why does it matter what is the function name in the MsgCall? Why does the ToAddress matter?
You are missing the point with this method, which is to generate a list of messages you can verify the filtering works, not that they are "valid" messages

Reply:
Since we talked internally, I've done a small refactor of the tests, adding some test cases and simplifying most of them. There is now a new test that tests the whole flow of the extractor - start to finish. [03fa2de](https://github.com/gnolang/tx-exports/pull/5/commits/03fa2de3dec655a4bed81cbb4c0469b9458e5a99), fixed its bugs in [3965302](https://github.com/gnolang/tx-exports/pull/5/commits/3965302f90c0f5773f9a0b05cadb3e66670b0028).